debian/clear_console.1
.\"***************************************************************************
.\" Copyright (c) 1998,2000 Free Software Foundation, Inc.                   *
.\"                                                                          *
.\" Permission is hereby granted, free of charge, to any person obtaining a  *
.\" copy of this software and associated documentation files (the            *
.\" "Software"), to deal in the Software without restriction, including      *
.\" without limitation the rights to use, copy, modify, merge, publish,      *
.\" distribute, distribute with modifications, sublicense, and/or sell       *
.\" copies of the Software, and to permit persons to whom the Software is    *
.\" furnished to do so, subject to the following conditions:                 *
.\"                                                                          *
.\" The above copyright notice and this permission notice shall be included  *
.\" in all copies or substantial portions of the Software.                   *
.\"                                                                          *
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS  *
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF               *
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.   *
.\" IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,   *
.\" DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR    *
.\" OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR    *
.\" THE USE OR OTHER DEALINGS IN THE SOFTWARE.                               *
.\"                                                                          *
.\" Except as contained in this notice, the name(s) of the above copyright   *
.\" holders shall not be used in advertising or otherwise to promote the     *
.\" sale, use or other dealings in this Software without prior written       *
.\" authorization.                                                           *
.\"***************************************************************************
.\"
.\" $Id: clear.1,v 1.3 2000/07/15 23:59:35 china Exp $
.TH clear_console 1 ""
.ds n 5
.SH NAME
\fBclear_console\fR - clear the console
.SH SYNOPSIS
\fBclear\fR
.br
.SH DESCRIPTION
\fBclear_console\fR clears your console if this is possible.  It looks in the
environment for the terminal type and then in the \fBterminfo\fR database to
figure out how to clear the screen. To clear the buffer, it then changes the
foreground virtual terminal to another terminal and then back to the original
terminal.
.SH SEE ALSO
\fBclear\fR(1), \fBchvt\fR(1)
.\"#
.\"# The following sets edit modes for GNU EMACS
.\"# Local Variables:
.\"# mode:nroff
.\"# fill-column:79
.\"# End:


debian/changelog
vyatta-bash (0.1) unstable; urgency=low
  
  * Initial release based on bash 3.1dfsg-8.

 -- An-Cheng Huang <ancheng@vyatta.com>  Mon, 12 Nov 2007 17:14:43 -0700

bash (3.1dfsg-8) unstable; urgency=high

  * bash-doc: Install copyright file when building without the
    gfdl'ed docs. Closes: #402471.

 -- Matthias Klose <doko@debian.org>  Mon, 11 Dec 2006 22:08:10 +0100

bash (3.1dfsg-7) unstable; urgency=medium

  * Restore the bash(1) manual page. Closes: #392989.

 -- Matthias Klose <doko@debian.org>  Sun, 22 Oct 2006 13:41:59 +0000

bash (3.1dfsg-6) unstable; urgency=low

  * Remove the texinfo documentation from the package. Closes: #357260.
  * Remove the FAQ from the package.

 -- Matthias Klose <doko@debian.org>  Thu, 12 Oct 2006 22:35:46 +0200

bash (3.1-5) unstable; urgency=low

  * Work around the broken remove-shell, not checking for the existance
    of /etc/shells (closes: #337097).
  * Force rebuild of bashref.info (closes: #354303, #357258).
  * Apply upstream patches 015, 016, 017, fixing
    - a problem with the extended globbing code prevented dots from matching
      filenames when used in some extended matching patterns.
    - core dumps when attempting to perform globbing in directories with
      very large numbers of files.
    - Array expansion failure with an arithmetic syntax error when the
      subscript appears within double quotes. Closes: #358831.
  * Add copyrights for contributed files (Joost van Baal). Closes: #357607.
  * bash_completion:
    - Add apt-cache madison completion: Closes: #342073, #359021.
    - Complete files with spaces for bzip2. Closes: #356218.
    - Complete whatis and apropos like man. Closes: #352145.
    - Ignore stderr in apt-cache completion. Closes: #367198.
    - Add flv|FLV files completion for mplayer. Closes: #364509, #373904.
    - Add dvi|DVI files completion for evince. Ubuntu #49880.
  * Fix typos in man page. Closes: #360543.

  [ Ian Jackson ]
  * Make bash die if it cannot read its input file.  Closes: #320036.
    (This is not an ideal fix because, for example, the exit status is
    still wrong - it should be 127, not 2.  Also, the filename is not
    printed.  Unfortunately the code is badly tangled making this small
    fix the most appropriate approach.)

 -- Matthias Klose <doko@debian.org>  Wed,  5 Jul 2006 06:56:54 +0200

bash (3.1-4) unstable; urgency=low

  * Update upstream patch 010, apply upstream patches 012, 013, 014.
  * bash-default-editor.dpatch: Fix typo (closes: #356308).
  * clear_console: Add FreeBSD support (Aurelian Jarno). Closes: #355336.
  * clear_console: Only get rid of any history in the scrollback, if
    not running in a pseudo tty (closes: #355815).
  * bash_completion:
    - Fix error in minicom completion (closes: #357241).

 -- Matthias Klose <doko@debian.org>  Thu, 23 Mar 2006 01:16:22 +0100

bash (3.1-3) unstable; urgency=low

  * Apply upstream patches 006, 007, 008, 009, 010 and 011.
  * Define PGRP_PIPE to avoid race condition; monitor mode causes emacs as
    $EDITOR to hang on kernel 2.6 (Jim Paris). Closes: #224543.
  * Fix read memleak when reading from non-blocking fd (Tim Waugh).
  * Recognize 'exec -l /bin/bash' as login shell (Tim Waugh).
  * Fix parameter expansion: Array member length expansion fails with nested
    index expression (Jan C. Nordholz). Closes: #345861.
  * Add swedish translation of message strings. Closes: #349677.

  Merge from Ubuntu:
  * clear_console: New helper program to clear the console, including
    the scrollback buffer.
  * /etc/skel/.bash_logout: Install it again and use clear_console.
    Ubuntu #29405. Closes: #331504.
  * /etc/skel/.bashrc: Enable bash_completion.
    Ubuntu #11414, #25096.
  * bash_completion:
    - Fix ssh completion to match lines starting with `Hostname' as well.
      Closes: #343312.
  * Add ulimit options -e and -r.

  * Update bash_completion to 20060301:
    - Completion for minicom(1), mtr(8), sysctl(8), smartctl(8), vncviewer(1),
      invoke-rc.d, update-rc.d and dpkg-source has been added.
      Closes: #323824, #327414.
    - gdb completion of second parameter was broken when first parameter
      contained white space.
    - gdb completion wasn't completing second parameter correctly when it was
      a file, rather than a PID.
    - Ruby ri completion has been broken for some time. This is now fixed.
    - Various fixes to work around change in how POSIX quoting is handled in
      bash 3.1.
    - subversion completion has been reimplemented from scratch and integrated
      into the main file.
    - iconv(1) completion has been improved.
    - yum(8) completion has been updated for current version of yum.
    - ant completion will now make use of complete-ant-cmd.pl, if available.
    - cvs(1) completion has been improved with 'update' and 'stat' completion.
    - 'aptitude show' now works in the same way as 'apt-cache show'.
    - make(1) now also completes on file names.
    - MPlayer will now also complete on .flac, .mpc and .3gp files.
      Closes: #340452.
    - wine will now also complete on .exe.so files.
    - unzip will now also complete on oowriter's .ott files.
    - xine et al will now complete on .mng files.
    - The list of programs completing on .dvi files has been expanded.
    - The range of files on which timidity and evince complete has been
      expanded. Closes: #351848.
    - mkisofs completion now defaults to treating results as file names.
    - $DEBUG has been renamed $BASH_COMPLETION_DEBUG to avoid namespace clashes
      with other software.
    - man(1) completion now works correctly on OpenBSD.
    - svk and Mercurial completion have been added to contribs. Closes: #324169.
    - Many other small optimisations and fixes. Closes: #325056.

 -- Matthias Klose <doko@debian.org>  Sat,  4 Mar 2006 01:54:25 +0100

bash (3.1-2) unstable; urgency=low

  * Apply upstream patches 002 - 005.

 -- Matthias Klose <doko@debian.org>  Tue, 10 Jan 2006 00:19:25 +0100


debian/clear_console.c
#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <unistd.h>
#include <sys/ioctl.h>
#if defined(__linux)
#include <linux/kd.h>
#include <linux/vt.h>
#elif defined(__FreeBSD_kernel__)
#include <sys/consio.h>
#include <sys/kbio.h>
#endif

#include <curses.h>
#include <term.h>

#define VERSION "0.1"

char* progname;
int quiet = 0;

void usage()
{
  fprintf(stderr, "Usage: %s [option]\n", progname);
  fprintf(stderr, "valid options are:\n");
  fprintf(stderr, "\t-q  --quiet    don't print error messages\n");
  fprintf(stderr, "\t-h --help      display this help text and exit\n");
  fprintf(stderr, "\t-V --version   display version information and exit\n");
}

const struct option opts[] =
{
  /* operations */
    {"help", no_argument, 0, 'h'},
    {"version", no_argument, 0, 'V'},
    {"quiet", no_argument, 0, 'q'},
    {0, 0, 0, 0}
};

static int putch(int c)
{
        return putchar(c);
}


/* taken from console-utils, lib/misc-console-utils.c */

int is_a_console(int fd)
{
#if defined(__linux__)
  char arg;
#elif defined(__FreeBSD_kernel__)
  int arg;
#endif

  arg = 0;
  return (ioctl(fd, KDGKBTYPE, &arg) == 0
          && ((arg == KB_OTHER) || (arg == KB_101) || (arg == KB_84)));
}

static int open_a_console(char *fnam)
{
  int fd;

  /* try read-only */
  fd = open(fnam, O_RDWR);

  /* if failed, try read-only */
  if (fd < 0 && errno == EACCES)
      fd = open(fnam, O_RDONLY);

  /* if failed, try write-only */
  if (fd < 0 && errno == EACCES)
      fd = open(fnam, O_WRONLY);

  /* if failed, fail */
  if (fd < 0)
      return -1;

  /* if not a console, fail */
  if (! is_a_console(fd))
    {
      close(fd);
      return -1;
    }

  /* success */
  return fd;
}

/*
 * Get an fd for use with kbd/console ioctls.
 * We try several things because opening /dev/console will fail
 * if someone else used X (which does a chown on /dev/console).
 *
 * if tty_name is non-NULL, try this one instead.
 */

int get_console_fd(char* tty_name)
{
  int fd;

  if (tty_name)
    {
      if (-1 == (fd = open_a_console(tty_name)))
        return -1;
      else
        return fd;
    }

  fd = open_a_console("/dev/tty");
  if (fd >= 0)
    return fd;

  fd = open_a_console("/dev/tty0");
  if (fd >= 0)
    return fd;

  fd = open_a_console("/dev/console");
  if (fd >= 0)
    return fd;

  for (fd = 0; fd < 3; fd++)
    if (is_a_console(fd))
      return fd;

#if 0
  fprintf(stderr,
          _("Couldnt get a file descriptor referring to the console\n"));
#endif
  return -1;            /* total failure */
}


int is_pseudo_tty(int fd)
{
  char *tty = ttyname(fd);

  if (!strncmp(tty, "/dev/pts/", 9))
    return 1;

  if (!strncmp(tty, "/dev/tty", 8) && tty[8] >= 'a' && tty[8] <= 'z')
    return 1;

  return 0;
}

int clear_console(int fd)
{
  int num, tmp_num;
#if defined(__linux__)
  struct vt_stat vtstat;
#endif

  /* clear screen */
  setupterm((char *) 0, 1, (int *) 0);
  if (tputs(clear_screen, lines > 0 ? lines : 1, putch) == ERR)
    {
      exit(1);
    }

  if (is_pseudo_tty(STDIN_FILENO))
    return 0;

  if (!strcmp(getenv("TERM"), "screen"))
      return 0;

  /* get current vt */
#if defined(__linux__)
  if (ioctl(fd, VT_GETSTATE, &vtstat) < 0)
#elif defined(__FreeBSD_kernel__)
  if (ioctl(fd, VT_ACTIVATE, &num) < 0)
#endif
    {
      if (!quiet)
	fprintf(stderr, "%s: cannot get VTstate\n", progname);
      exit(1);
    }
#if defined(__linux__)
  num = vtstat.v_active;
#endif
  tmp_num = (num == 1 ? 2 : 1);

  /* switch vt to clear the scrollback buffer */
  if (ioctl(fd, VT_ACTIVATE, tmp_num))
    {
      if (!quiet)
	perror("chvt: VT_ACTIVATE");
      exit(1);
    }

  if (ioctl(fd, VT_WAITACTIVE, tmp_num))
    {
      if (!quiet)
	perror("VT_WAITACTIVE");
      exit(1);
    }

  /* switch back */
  if (ioctl(fd, VT_ACTIVATE, num))
    {
      if (!quiet)
	perror("chvt: VT_ACTIVATE");
      exit(1);
    }

  if (ioctl(fd, VT_WAITACTIVE, num))
    {
      if (!quiet)
	perror("VT_WAITACTIVE");
      exit(1);
    }
}

int main (int argc, char* argv[])
{
  int fd;
  int result;                          /* option handling */
  int an_option;

  if ((progname = strrchr(argv[0], '/')) == NULL)
    progname = argv[0];
  else
    progname++;

  while (1)
    {
      result = getopt_long(argc, argv, "Vhq", opts, &an_option);
      
      if (result == EOF)
	break;

      switch (result)
        {
        case 'V':
	  fprintf(stdout, "%s: Version %s\n", progname, VERSION);
          exit (0);
        case 'h':
          usage();
          exit (0);
	  
        case 'q':
          quiet = 1;
        }
    }

  if (optind < argc)
    {
      if (!quiet)
	fprintf(stderr, "%s: no non-option arguments are valid", progname);
      exit(1);
    }

  if ((fd = get_console_fd(NULL)) == -1)
    {
      if (!quiet)
	fprintf(stderr, "%s: terminal is not a console\n", progname);
      exit(1);
    }

  clear_console(fd);

  return 0;
}
